contracts/near/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+    SetGovernanceSource(Source),
+    SetStalePriceThreshold(u64),
+    SetUpdateFee(Balance),
+}
+
+#[near_bindgen]
+impl Pyth {
+    #[init]
+    #[allow(clippy::new_without_default)]
+    pub fn new(
+        wormhole: AccountId,
+        codehash: [u8; 32],
+        initial_source: Source,
+        gov_source: Source,
+        update_fee: Balance,
+        stale_threshold: u64,
+    ) -> Self {
+        // Add an initial Source so that the contract can be used.
+        let mut sources = UnorderedSet::new(StorageKeys::Source);
+        sources.insert(&initial_source);
+        Self {
+            prices: UnorderedMap::new(StorageKeys::Prices),
+            last_gov_seq: 0,
+            stale_threshold,
+            gov_source,
+            sources,
+            wormhole,
+            codehash,
+            update_fee,
+        }
+    }
+
+    #[init(ignore_state)]
+    pub fn migrate() -> Self {
+        let state: Self = env::state_read().expect("Failed to read state");
+        state
+    }
+
+    /// Instruction for processing VAA's relayed via Wormhole.
+    ///
+    /// Note that VAA verification requires calling Wormhole so processing of the VAA itself is
+    /// done in a callback handler, see `process_vaa_callback`.
+    #[payable]
+    #[handle_result]
+    pub fn update_price_feeds(&mut self, vaas: Vec<String>) -> Result<(), Error> {
+        for vaa_hex in vaas {
+            // We Verify the VAA is coming from a trusted source chain before attempting to verify
+            // VAA signatures. Avoids a cross-contract call early.
+            let vaa = hex::decode(&vaa_hex).map_err(|_| Error::InvalidHex)?;
+            let vaa = vaa_payload::from_slice_with_payload::<structs::Vaa<()>>(&vaa);
+            let vaa = vaa.map_err(|_| Error::InvalidVaa)?;
+            let (vaa, _rest) = vaa;
+
+            // Convert to local VAA type to catch APi changes.
+            let vaa = Vaa::from(vaa);
+
+            if self.sources.contains(&Source {
+                emitter:            vaa.emitter_address,
+                pyth_emitter_chain: vaa.emitter_chain,
+            }) {
+                return Err(Error::UnknownSource);
+            }
+
+            // Verify VAA and refund the caller in case of failure.
+            ext_wormhole::ext(env::current_account_id())
+                .with_static_gas(Gas(30_000_000_000_000))
+                .verify_vaa(vaa_hex.clone())
+                .then(
+                    Self::ext(env::current_account_id())
+                        .with_static_gas(Gas(30_000_000_000_000))
+                        .with_attached_deposit(env::attached_deposit())
+                        .verify_vaa_callback(env::predecessor_account_id(), vaa_hex),
+                )
+                .then(
+                    Self::ext(env::current_account_id())
+                        .with_static_gas(Gas(30_000_000_000_000))
+                        .refund_vaa(env::predecessor_account_id(), env::attached_deposit()),
```

Comment:
Since we are in a loop, are you sure the total refund works correctly? :think: refunding works ok, but refunding extra deposit probably not:
```
            // Otherwise we refund whatever is left over.
            if deposit - cost > 0 {
                Promise::new(refunder).transfer(cost);
            }
```

Reply:
As above.